File: examples/mastra/console/warsaw-example.ts
/**
 * Warsaw, Poland - Mapbox MCP Server Demo (Mastra)
 *
 * This example demonstrates the Mapbox MCP server with Mastra using
 * landmarks and locations in Warsaw, Poland. Perfect for demos and talks!
 *
 * Learn more about Mastra: https://mastra.ai
 */

import { Agent } from '@mastra/core';
import { MCPClient } from '@mastra/mcp';

// Configure Mapbox MCP Client
export const mapboxMcpClient = new MCPClient({
  id: 'mapbox-mcp-client',
  servers: {
    mapbox: {
      // Use local build from repository
      command: 'node',
      args: ['../../dist/esm/index.js'],
      env: {
        MAPBOX_ACCESS_TOKEN: process.env.MAPBOX_ACCESS_TOKEN || ''
      }
    }
  }
});

/**
 * Create an agent specialized for Warsaw tourism
 */
async function createWarsawAgent() {
  const tools = await mapboxMcpClient.getTools({ serverName: 'mapbox' });

  return new Agent({
    name: 'Warsaw Tourism Expert',
    instructions: `You are an expert guide for Warsaw, Poland with access to Mapbox geospatial tools.
    You help visitors discover famous landmarks, plan routes, and find interesting places in Warsaw.
    Provide clear, helpful information about locations, distances, and travel times.`,
    model: {
      provider: 'OPEN_AI',
      name: 'gpt-4o-mini',
      toolChoice: 'auto'
    },
    tools: tools
  });
}

/**
 * Find a Warsaw landmark
 */
async function findLandmark(agent: Agent, landmarkName: string) {
  console.log('\n' + '='.repeat(70));
  console.log(`🔍 Finding: ${landmarkName}`);
  console.log('='.repeat(70) + '\n');

  const result = await agent.generate(
    `Find ${landmarkName} in Warsaw, Poland. Provide the coordinates and address.`
  );

  console.log(result.text);
  return result;
}

/**
 * Get directions between two locations
 */
async function getDirections(
  agent: Agent,
  origin: string,
  destination: string
) {
  console.log('\n' + '='.repeat(70));
  console.log(`🗺️  Route: ${origin} → ${destination}`);
  console.log('='.repeat(70) + '\n');

  const result = await agent.generate(
    `Get walking directions from ${origin} to ${destination} in Warsaw, Poland. ` +
      `Include the distance and estimated travel time.`
  );

  console.log(result.text);
  return result;
}

/**
 * Find places near a location
 */
async function findNearbyPlaces(
  agent: Agent,
  location: string,
  category: string
) {
  console.log('\n' + '='.repeat(70));
  console.log(`🔎 Finding ${category} near ${location}`);
  console.log('='.repeat(70) + '\n');

  const result = await agent.generate(
    `Find 3-5 ${category} near ${location} in Warsaw, Poland. ` +
      `Include names and brief descriptions.`
  );

  console.log(result.text);
  return result;
}

/**
 * Main demo function
 */
async function main() {
  console.log('='.repeat(70));
  console.log('🇵🇱 Warsaw, Poland - Mapbox MCP Server Demo (Mastra)');
  console.log('='.repeat(70));

  try {
    const agent = await createWarsawAgent();

    // Scenario 1: Find famous Warsaw landmarks
    console.log('\n' + '='.repeat(70));
    console.log('SCENARIO 1: Famous Warsaw Landmarks');
    console.log('='.repeat(70));

    const landmarks = [
      'Palace of Culture and Science', // Pałac Kultury i Nauki
      'Royal Castle', // Zamek Królewski
      'Old Town Market Square' // Rynek Starego Miasta
    ];

    for (const landmark of landmarks) {
      await findLandmark(agent, landmark);
      await new Promise((resolve) => setTimeout(resolve, 1000)); // Brief pause
    }

    // Scenario 2: Get directions between landmarks
    console.log('\n' + '='.repeat(70));
    console.log('SCENARIO 2: Navigate Between Landmarks');
    console.log('='.repeat(70));

    await getDirections(agent, 'Old Town Market Square', 'Royal Castle');

    // Scenario 3: Find nearby places
    console.log('\n' + '='.repeat(70));
    console.log('SCENARIO 3: Discover Nearby Places');
    console.log('='.repeat(70));

    await findNearbyPlaces(agent, 'Palace of Culture and Science', 'cafes');

    // Scenario 4: Find more specific locations
    console.log('\n' + '='.repeat(70));
    console.log('SCENARIO 4: Additional Warsaw Locations');
    console.log('='.repeat(70));

    const otherLocations = [
      'Lazienki Park', // Łazienki Park
      'Wilanow Palace', // Pałac w Wilanowie
      'Warsaw Uprising Museum'
    ];

    for (const location of otherLocations) {
      await findLandmark(agent, location);
      await new Promise((resolve) => setTimeout(resolve, 1000));
    }

    console.log('\n' + '='.repeat(70));
    console.log('✅ Demo Complete!');
    console.log('='.repeat(70));
    console.log('\nThese examples demonstrate:');
    console.log('• Location search and geocoding');
    console.log('• Turn-by-turn directions');
    console.log('• Point of interest discovery');
    console.log('• TypeScript-first agent development');
    console.log('• MCP integration with Mastra');
  } finally {
    console.log('\nDisconnecting from MCP servers...');
    await mapboxMcpClient.disconnect();
    console.log('Disconnected.');
  }
}

// Run the demo
main().catch(console.error);

File: examples/pydantic-ai/warsaw_example.py
"""
Warsaw, Poland - Mapbox MCP Server Demo

This example demonstrates the Mapbox MCP server with Pydantic AI using
landmarks and locations in Warsaw, Poland. Perfect for demos and talks!

Learn more about Pydantic AI: https://ai.pydantic.dev
"""

import asyncio
import os
from dotenv import load_dotenv
from pydantic import BaseModel
from pydantic_ai import Agent
from pydantic_ai.mcp import MCPServerStdio
from pydantic_ai.models.openai import OpenAIChatModel

# Load environment variables from .env file
load_dotenv()


# Pydantic model for structured location output
class Location(BaseModel):
    """Structured location data with coordinates and details."""
    name: str
    latitude: float
    longitude: float
    address: str | None = None
    description: str | None = None


def get_mcp_server():
    """Configure the Mapbox MCP server connection."""
    mapbox_token = os.environ.get("MAPBOX_ACCESS_TOKEN")
    if not mapbox_token:
        raise EnvironmentError(
            "MAPBOX_ACCESS_TOKEN environment variable is required. "
            "Get your token at https://account.mapbox.com/"
        )

    # Use local build from repository
    script_dir = os.path.dirname(os.path.abspath(__file__))
    script_path = os.path.join(script_dir, "..", "..", "dist", "esm", "index.js")

    return MCPServerStdio(
        "node",
        args=[script_path],
        env={"MAPBOX_ACCESS_TOKEN": mapbox_token},
        timeout=30
    )


async def find_landmark(landmark_name: str) -> Location:
    """
    Find a Warsaw landmark and return structured location data.

    Args:
        landmark_name: Name of the landmark to find

    Returns:
        Location object with coordinates and details
    """
    print(f"\n{'='*70}")
    print(f"🔍 Finding: {landmark_name}")
    print(f"{'='*70}\n")

    mcp_server = get_mcp_server()

    agent = Agent(
        OpenAIChatModel("gpt-4o-mini"),
        output_type=Location,
        system_prompt="""You are a Warsaw tourism expert with access to Mapbox tools.
        Find accurate location information and return structured data.""",
        toolsets=[mcp_server]
    )

    result = await agent.run(f"Find {landmark_name} in Warsaw, Poland")
    location = result.output

    print(f"📍 {location.name}")
    print(f"   Coordinates: {location.latitude:.6f}, {location.longitude:.6f}")
    if location.address:
        print(f"   Address: {location.address}")
    if location.description:
        print(f"   {location.description}")

    return location


async def get_directions(origin: str, destination: str) -> str:
    """
    Get directions between two locations in Warsaw.

    Args:
        origin: Starting location
        destination: Ending location

    Returns:
        Directions as text
    """
    print(f"\n{'='*70}")
    print(f"🗺️  Route: {origin} → {destination}")
    print(f"{'='*70}\n")

    mcp_server = get_mcp_server()

    agent = Agent(
        OpenAIChatModel("gpt-4o-mini"),
        system_prompt="""You are a navigation expert with access to Mapbox tools.
        Provide clear, helpful directions including travel time and distance.""",
        toolsets=[mcp_server]
    )

    result = await agent.run(
        f"Get walking directions from {origin} to {destination} in Warsaw, Poland. "
        f"Include the distance and estimated travel time."
    )

    print(result.output)
    return result.output


async def find_nearby_places(location: str, category: str) -> str:
    """
    Find places near a location in Warsaw.

    Args:
        location: The location to search near
        category: Type of place to find (e.g., "restaurants", "cafes", "museums")

    Returns:
        List of places as text
    """
    print(f"\n{'='*70}")
    print(f"🔎 Finding {category} near {location}")
    print(f"{'='*70}\n")

    mcp_server = get_mcp_server()

    agent = Agent(
        OpenAIChatModel("gpt-4o-mini"),
        system_prompt="""You are a local Warsaw guide with access to Mapbox tools.
        Find interesting places and provide helpful recommendations.""",
        toolsets=[mcp_server]
    )

    result = await agent.run(
        f"Find 3-5 {category} near {location} in Warsaw, Poland. "
        f"Include names and brief descriptions."
    )

    print(result.output)
    return result.output


async def main():
    """
    Demo scenarios for Warsaw, Poland.

    This showcases various Mapbox MCP capabilities using Warsaw landmarks
    that would be familiar to a Polish audience.
    """
    print("="*70)
    print("🇵🇱 Warsaw, Poland - Mapbox MCP Server Demo")
    print("="*70)

    # Scenario 1: Find famous Warsaw landmarks
    print("\n" + "="*70)
    print("SCENARIO 1: Famous Warsaw Landmarks")
    print("="*70)

    landmarks = [
        "Palace of Culture and Science",  # Pałac Kultury i Nauki
        "Royal Castle",                    # Zamek Królewski
        "Old Town Market Square"           # Rynek Starego Miasta
    ]

    landmark_locations = {}
    for landmark in landmarks:
        location = await find_landmark(landmark)
        landmark_locations[landmark] = location
        await asyncio.sleep(1)  # Brief pause between requests

    # Scenario 2: Get directions between landmarks
    print("\n" + "="*70)
    print("SCENARIO 2: Navigate Between Landmarks")
    print("="*70)

    await get_directions(
        "Old Town Market Square",
        "Royal Castle"
    )

    # Scenario 3: Find nearby places
    print("\n" + "="*70)
    print("SCENARIO 3: Discover Nearby Places")
    print("="*70)

    await find_nearby_places(
        "Palace of Culture and Science",
        "cafes"
    )

    # Scenario 4: Find more specific locations
    print("\n" + "="*70)
    print("SCENARIO 4: Additional Warsaw Locations")
    print("="*70)

    other_locations = [
        "Lazienki Park",        # Łazienki Park
        "Wilanow Palace",       # Pałac w Wilanowie
        "Warsaw Uprising Museum"
    ]

    for location in other_locations:
        await find_landmark(location)
        await asyncio.sleep(1)

    print("\n" + "="*70)
    print("✅ Demo Complete!")
    print("="*70)
    print("\nThese examples demonstrate:")
    print("• Location search and geocoding")
    print("• Structured output with Pydantic models")
    print("• Turn-by-turn directions")
    print("• Point of interest discovery")
    print("• Type-safe AI agents with MCP integration")


if __name__ == "__main__":
    asyncio.run(main())
